### PR TITLE
[SYCL][ESIMD] Convert esimd operator to friend method.

### PR DESCRIPTION
**Pull request by cmc-rep**
Signed-off-by: Gang Y Chen <gang.y.chen@intel.com>

### CODE REVIEW

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp [hunk 1]
```diff
@@ -184,30 +184,26 @@ template <typename Ty, int N> class simd {
   // {/quote}
   //   * if not different, then auto should not be used
 #define DEF_BINOP(BINOP, OPASSIGN)                                             \
-  auto operator BINOP(const simd &RHS) const {                                 \
+  ESIMD_INLINE friend auto operator BINOP(const simd &x, const simd &y) {      \
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:187]: Please make parameter naming consistent with other friends (here and below):
```suggestion
  ESIMD_INLINE friend auto operator BINOP(const simd &Y, const simd &Y) {      \
```

**Comment by cmc-rep** [comment]
I am waiting for the test result. It seems taking forever

**Review by vladimirlaz — DISMISSED** [review]
Approve to start testing

**Comment by cmc-rep** [comment]
Updated change

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp [hunk 2]
```diff
@@ -222,8 +218,9 @@ template <typename Ty, int N> class simd {
   // the future revisions.
   //
 #define DEF_RELOP(RELOP)                                                       \
-  simd<uint16_t, N> operator RELOP(const simd &RHS) const {                    \
-    auto R = data() RELOP RHS.data();                                          \
+  ESIMD_INLINE friend simd<uint16_t, N> operator RELOP(const simd &X,          \
+                                                       const simd &Y) {        \
+    auto R = X.data() RELOP Y.data();                                          \
     mask_type_t<N> M(1);                                                       \
```
[the remark this answers is not in the document]
Comment (cmc-rep) [reply sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:224]: M(1) is not part of my change. Not sure why it is written this way. I would be cautious to make unnecessary change

File: sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp [hunk 3]
```diff
@@ -180,13 +180,26 @@ template <typename BaseTy, typename RegionTy> class simd_view {
...
+    auto V0 = convert<typename ComputeTy::vector_type>(X.data());              \
+    auto V1 = convert<typename ComputeTy::vector_type>(Y.read().data());       \
     auto V2 = V0 BINOP V1;                                                     \
     return ComputeTy(V2);                                                      \
   }                                                                            \
+  ESIMD_INLINE friend auto operator BINOP(const simd_view &X,                  \
+                                          const simd_view &Y) {                \
+    return (X BINOP Y.read());                                                 \
+  }                                                                            \
   simd_view &operator OPASSIGN(const value_type &RHS) {                        \
```
[the remark this answers is not in the document]
Comment (cmc-rep) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp:203]: Cannot have that, cause ambiguous with the friend method definition in simd class

File: sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp [hunk 4]
```diff
@@ -205,10 +221,21 @@ template <typename BaseTy, typename RegionTy> class simd_view {
...
+  ESIMD_INLINE friend simd<uint16_t, length> operator RELOP(                   \
+      const value_type &X, const simd_view &Y) {                               \
+    auto R = X.data() RELOP Y.read().data();                                   \
+    mask_type_t<length> M(1);                                                  \
+    return M & convert<mask_type_t<length>>(R);                                \
+  }                                                                            \
+  ESIMD_INLINE friend simd<uint16_t, length> operator RELOP(                   \
+      const simd_view &X, const simd_view &Y) {                                \
+    return (X RELOP Y.read());                                                 \
   }
```
[the remark this answers is not in the document]
Comment (cmc-rep) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp:239]: Cannot have that, cause ambiguous with the friend method definition in simd class

File: sycl/include/CL/sycl/INTEL/esimd/esimd.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/esimd.hpp:224]: Yes, I see that it is not part of this change, that's why it is just a nit for another patch. I thought you wrote this part.

File: sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp:203]: I see thanks.

**Review by kbobrovs — APPROVED** [review]

**Comment by cmc-rep** [comment]
@kbobrovs please help me merge this PR.  Somehow I am not authorized to merge

**Review by kychendev — DISMISSED** [review]

File: sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp [hunk 5]
```diff
@@ -389,7 +429,7 @@ template <typename BaseTy, typename RegionTy> class simd_view {
   // - std::pair<top_region_type, base_region_type>
   //
   RegionTy M_region;
-};
+}; // namespace gpu
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp:432]: Missed this one in the initial review. Please fix.
```suggestion
};
```

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp (hunk 5 above)
Comment (cmc-rep) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp:432]: I will work on some change to the same file again. Could I fix it in the next change. So I can move on working on the next change

Comment (cmc-rep) [reply sycl/include/CL/sycl/INTEL/esimd/esimd_view.hpp]: @kbobrovs, fix it. Please review and merge timely.

**Review by kbobrovs — APPROVED** [review]

**Review by kychendev — APPROVED** [review]

**Comment by cmc-rep** [comment]
@Konst, this PR is ready, please help with merge

**Comment by cmc-rep** [comment]
@kbobrovs this PR is ready, please help with merge

**Pull request merged**